src/databases/NASTRAN/avtNASTRANFileFormat.C

```diff
@@ -187,90 +187,75 @@ avtNASTRANFileFormat::ActivateTimestep()
-
-static float Getf(const char *s)
+static double GetVal(const char *s)
 {
-    char *ends;
-    double val = 0.0;
+    double dval;
+    char *endptr = 0;
+    size_t i;
+
+    // handle a string with no numbers
+    i = strcspn(s, "0123456789");
+    if (s[i] == '\0') return 0;
 
-    // Check for one of these funky 'NASTRAN exponential format' strings.
+    // Check for the funky 'NASTRAN exponential format'.
     // This is where a value like '1.2345e-5' is actually represented in the
     // file as '1.2345-5' with the 'e' character missing. It is awkward but
     // apparently a NASTRAN standard. I guess the rationale is that given
     // an 8 character field width limit, removing the 'e' character gives them
     // one additional digit of precision. This logic is basically looking for
     // the condition of encountering a sign character, '-' or '+', AFTER having
     // seen characters that could represent part of a number. In such a case,
-    // it MUST be the sign of the exponent.
-    const char *p = s;
-    char tmps[32];
-    char *q = tmps;
-    bool haveSeenNumChars = false;
-    while (!haveSeenNumChars || (*p != '-' && *p != '+' && *p != '\0'))
-    {
-        if (('0' <= *p && *p <= '9') || *p == '.' || *p == '+' || *p == '-')
-            haveSeenNumChars = true;
-        *q++ = *p++;
-    }
-    if (haveSeenNumChars && (*p == '-' || *p == '+'))
-    {
-        *q++ = 'e';
-        while (*p != '\0')
-            *q++ = *p++;
-        *q++ = '\0';
-        errno = 0;
-        val = strtod(tmps, &ends);
-    }
-    else
-    {
-        errno = 0;
-        val = strtod(s, &ends);
-    }
+    // it MUST be the sign of an exponent.
+    std::string t(s);
+    for (i = t.size(); t[i]!='-' && t[i]!='+' && t[i]!='\0'; i--);
+    if (i>0 && (t[i]=='-' || t[i]=='+') && t[i-1]!='e' && t[i-1]!='E')
+        t.insert(i, "E");
+
+    errno = 0;
+    dval = strtod(t.c_str(), &endptr);
 
-    if (errno != 0)
+    // Handle cases where conversion failed for one reason or another
+    if ((dval == 0 && endptr == t.c_str()) || errno)
     {
-        char msg[512];
-        snprintf(msg, sizeof(msg),
-            "Error \"%s\" at word \"%32s\"\n", strerror(errno), s);
-        if (!avtCallback::IssueWarning(msg))
-            cerr << msg << endl;
-        return 0.0;
+        static int nwarnings = 0;
+        if (nwarnings <= 5) 
+        {
+            char msg[512];
+            snprintf(msg, sizeof(msg), "%s problem converting numerical value from \"%32s\"\n",
+                errno?strerror(errno):"", s);
+            if (!avtCallback::IssueWarning(msg))
+            {
+                cerr << msg << endl;
+                debug5 << msg << endl;
+            }
+            if (nwarnings == 5)
+            {
+                if (!avtCallback::IssueWarning("Further warnings will be supressed."))
+                {
+                    cerr << "Further warnings will be supressed." << endl;
```

Comment:
Should output to cerr be removed?

Reply:
No...at most 5 lines of output will appear there *only* if it was unable to issue the warning message in the call on line 236.